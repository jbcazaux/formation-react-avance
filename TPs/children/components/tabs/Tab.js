import PropTypes from 'prop-types'

const Tab = ({ children }) => children

Tab.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node,
}

export default Tab
