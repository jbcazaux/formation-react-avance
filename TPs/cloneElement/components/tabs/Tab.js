import React from 'react'
import PropTypes from 'prop-types'

// TODO : utiliser setDetail
const Tab = ({ children, setDetail }) => children

Tab.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node,
  setDetail: PropTypes.func,
}

export default Tab
