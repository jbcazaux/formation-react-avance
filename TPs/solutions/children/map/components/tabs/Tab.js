import PropTypes from 'prop-types'

const Tab = () => null

Tab.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node,
}

export default Tab
