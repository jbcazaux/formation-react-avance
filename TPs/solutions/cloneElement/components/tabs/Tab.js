import React from 'react'
import PropTypes from 'prop-types'

const Tab = () => null

Tab.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node,
  setDetail: PropTypes.func,
}

export default Tab
