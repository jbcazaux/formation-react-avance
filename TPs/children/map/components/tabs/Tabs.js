import React, { useState } from 'react'
import style from './Tabs.scss'
import PropTypes from 'prop-types'
import Tab from 'components/tabs/Tab'
import cn from 'classnames'

const Tabs = ({ children }) => {
  // TODO 1: Créer un state pour sauvegarder l'index de l'onglet sélectionné

  return (
    <div>
      <div className={style.tabs}>
        {/* Affichage des titres des onglets */}
        {React.Children.map(children, (child, index) => {
          if (child.type !== Tab) return null
          return (
            // TODO 2: <TabTitle onClick={mettre à jour l index sélectionné} selected={?}>TITRE</TabTitle>
            null
          )
        })}
      </div>
      {/* Affichage de l'onglet actif (c'est à dire un des <Tab /> ) */}
      {React.Children.map(children, child => {
        if (child.type !== Tab) return null
        // TODO 3 : Ne pas afficher les onglets non actifs (ceux qui n'ont pas le bon index)
        return null // TODO 4 : Retourner l'onglet, c'est à dire l'élément sur lequel on itère
      })}
    </div>
  )
}

Tabs.propTypes = {
  children: PropTypes.arrayOf(PropTypes.node).isRequired,
}

export default Tabs

const TabTitle = ({ selected, onClick }) => (
  <div className={cn(style.tab, { [style.selectedTab]: selected })} onClick={onClick}>
    TITRE {/* TODO 5: comment récuperer le titre de la props title du composant <Tab />? */}
  </div>
)

TabTitle.propTypes = {
  selected: PropTypes.bool,
  onClick: PropTypes.func,
}
