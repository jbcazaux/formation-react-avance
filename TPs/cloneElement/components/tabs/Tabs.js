import React, { useState } from 'react'
import style from './Tabs.scss'
import PropTypes from 'prop-types'
import Tab from './Tab'
import cn from 'classnames'

const Tabs = ({ children }) => {
  const [selectedTab, setSelectedTab] = useState(0)
  const [details, setDetails] = useState({})

  return (
    <div role="tabs">
      <div className={style.tabs} role="tablist">
        {React.Children.map(children, (child, index) => {
          if (child.type !== Tab) return null
          const onclick = setSelectedTab.bind(null, index)
          return (
            <TabTitle selected={selectedTab === index} onClick={onclick} detail={details[index]}>
              {child.props.title}
            </TabTitle>
          )
        })}
      </div>
      {React.Children.map(children, (child, index) => {
        if (index !== selectedTab) return null
        if (child.type !== Tab) return null
        const setDetail = d => setDetails(prev => ({ ...prev, [index]: d }))
        // TODO 1: passer setDetail au 'child' (child est un Tab, cf App.js)
        return child
      })}
    </div>
  )
}

Tabs.propTypes = {
  children: PropTypes.arrayOf(PropTypes.node).isRequired,
}

export default Tabs

const TabTitle = ({ selected, children, detail, onClick }) => (
  <div className={cn(style.tab, { [style.selectedTab]: selected })} onClick={onClick} role="tab">
    {children}
    <div className={style.detail} role="note">
      {detail}&#8203;
    </div>
  </div>
)

TabTitle.propTypes = {
  detail: PropTypes.string,
  children: PropTypes.node.isRequired,
  selected: PropTypes.bool,
  onClick: PropTypes.func,
}
