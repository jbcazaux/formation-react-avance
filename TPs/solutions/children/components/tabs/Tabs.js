import React, { useState } from 'react'
import style from './Tabs.scss'
import PropTypes from 'prop-types'
import Tab from 'components/tabs/Tab'
import cn from 'classnames'

const Tabs = ({ children }) => {
  const [selectedTab, setSelectedTab] = useState(0)

  return (
    <div>
      <div className={style.tabs}>
        {React.Children.map(children, (child, index) => {
          if (child.type !== Tab) return null
          const onclick = setSelectedTab.bind(null, index)
          return (
            <TabTitle selected={selectedTab === index} onClick={onclick}>
              {child.props.title}
            </TabTitle>
          )
        })}
      </div>
      {React.Children.map(children, (child, index) => {
        if (index !== selectedTab) return null
        if (child.type !== Tab) return null
        return child
      })}
    </div>
  )
}

Tabs.propTypes = {
  children: PropTypes.arrayOf(PropTypes.node).isRequired,
  defaultInfos: PropTypes.object,
}

export default Tabs

const TabTitle = ({ selected, children, onClick }) => (
  <div className={cn(style.tab, { [style.selectedTab]: selected })} onClick={onClick}>
    {children}
  </div>
)

TabTitle.propTypes = {
  children: PropTypes.node.isRequired,
  selected: PropTypes.bool,
  onClick: PropTypes.func,
}
