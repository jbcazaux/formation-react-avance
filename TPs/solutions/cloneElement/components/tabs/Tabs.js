import React, { useState } from 'react'
import style from './Tabs.scss'
import PropTypes from 'prop-types'
import Tab from 'components/tabs/Tab'
import cn from 'classnames'

const Tabs = ({ children }) => {
  const [selectedTab, setSelectedTab] = useState(0)
  const [sousTitres, setSousTitres] = useState({})

  return (
    <div>
      <div className={style.tabs}>
        {React.Children.map(children, (child, index) => {
          if (child.type !== Tab) return null
          const onclick = setSelectedTab.bind(null, index)
          return (
            <TabTitle selected={selectedTab === index} onClick={onclick} subTitle={sousTitres[index]}>
              {child.props.title}
            </TabTitle>
          )
        })}
      </div>
      {React.Children.map(children, (child, index) => {
        if (index !== selectedTab) return null
        if (child.type !== Tab) return null
        const setSousTitre = st => setSousTitres(prev => ({ ...prev, [index]: st }))
        return React.cloneElement(child.props.children, { setSousTitre })
      })}
    </div>
  )
}

Tabs.propTypes = {
  children: PropTypes.arrayOf(PropTypes.node).isRequired,
  defaultInfos: PropTypes.object,
}

export default Tabs

const TabTitle = ({ selected, children, subTitle, onClick }) => (
  <div className={cn(style.tab, { [style.selectedTab]: selected })} onClick={onClick}>
    {children}
    <div className={style.soustitre}>{subTitle}&#8203;</div>
  </div>
)

TabTitle.propTypes = {
  subTitle: PropTypes.string,
  children: PropTypes.node.isRequired,
  selected: PropTypes.bool,
  onClick: PropTypes.func,
}
