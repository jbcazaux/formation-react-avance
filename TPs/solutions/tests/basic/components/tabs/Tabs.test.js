import React from 'react'
import { render, fireEvent, getNodeText, screen } from '@testing-library/react'
import Tabs from './Tabs'
import Tab from './Tab'

describe('Tabs', () => {
  const Body1 = ({ setDetail }) => (
    <div onClick={() => setDetail('info !')} data-testid="body1">
      body1
    </div>
  )
  const Body2 = () => <div data-testid="body2">body2</div>

  it('displays simple Tabs', () => {
    const { container } = render(
      <Tabs>
        <Tab title="title1">
          <Body1 />
        </Tab>
        <Tab title="title2">
          <Body2 />
        </Tab>
      </Tabs>
    )

    expect(container).toMatchSnapshot()
  })

  it('displays second tab body when clicking on title2', async () => {
    render(
      <Tabs>
        <Tab title="title1">
          <Body1 />
        </Tab>
        <Tab title="title2">
          <Body2 />
        </Tab>
      </Tabs>
    )
    expect(screen.queryByTestId('body1')).toBeTruthy()
    expect(screen.queryByTestId('body2')).toBeFalsy()

    const secondTab = screen.getAllByRole('tab')[1]
    fireEvent.click(secondTab)

    expect(screen.queryByTestId('body1')).toBeFalsy()
    expect(screen.queryByTestId('body2')).toBeTruthy()
  })

  it('displays detail after setting it', () => {
    render(
      <Tabs>
        <Tab title="title1">
          <Body1 />
        </Tab>
        <Tab title="title2">
          <Body2 />
        </Tab>
      </Tabs>
    )

    const firstTabDetails = screen.getAllByRole('note')[0]
    expect(getNodeText(firstTabDetails)).not.toContain('info !')
    fireEvent.click(screen.getByTestId('body1'))
    expect(getNodeText(firstTabDetails)).toContain('info !')
  })
})
