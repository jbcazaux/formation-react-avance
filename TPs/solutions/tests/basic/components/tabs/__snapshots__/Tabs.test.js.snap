// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs displays simple Tabs 1`] = `
<div>
  <div>
    <div
      class="tabs"
    >
      <div
        class="tab selectedTab"
      >
        title1
        <div
          class="soustitre"
        >
          ​
        </div>
      </div>
      <div
        class="tab"
      >
        title2
        <div
          class="soustitre"
        >
          ​
        </div>
      </div>
    </div>
    <div
      data-testid="body1"
    >
      body1
    </div>
  </div>
</div>
`;
